### ISSUE TITLE
Bug: [no-redundant-type-constituents] Differentiate a types-error any from a "true" any

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#9214]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.5&fileType=.ts&code=C4TwDgpgBAogTnA9nAKuCBnKBeKA5RYAaQDtEB3EqAHygAYBuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1ukQBNYmvAIYtaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=&tokens=false

### Repro Code

```TypeScript
type ErrorTypes = NotKnown | 0;
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-redundant-type-constituents": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

The error message shouldn't phrase the `NotKnown` as "`any`", since it's really an "error" type. It's confusing to have it mentioned as `any`.

### Actual Result

```plaintext
'any' overrides all other types in this union type. 1:19 - 1:27
```

### Additional Info

Split out of #8304 <- #8231.

💖
